### PR TITLE
select related field paths

### PR DESCRIPTION
**Pull request by squarethecircle**

### CODE REVIEW

File: mongoengine/dereference.py [hunk 1]
```diff
@@ -89,6 +131,9 @@ def _find_references(self, items, depth=0):
         if not items or depth >= self.max_depth:
             return reference_map
 
+        if field_paths is not None and not isinstance(field_paths, dict):
+            return reference_map
```
Comment (myprasanna) [inline mongoengine/dereference.py:135]: why this extra protection? can we remove this?

File: mongoengine/dereference.py [hunk 2]
```diff
@@ -100,33 +145,40 @@ def _find_references(self, items, depth=0):
...
                 reference_map.setdefault(get_document(item['_cls']), set()).add(item['_ref'].id)
             elif isinstance(item, (dict, list, tuple)) and depth - 1 <= self.max_depth:
-                references = self._find_references(item, depth - 1)
+                references = self._find_references(item, depth - 1, field_paths=field_paths)
                 for key, refs in iteritems(references):
                     reference_map.setdefault(key, set()).update(refs)
-
         return reference_map
 
     def _fetch_objects(self, doc_type=None):
```
Comment (myprasanna) [inline mongoengine/dereference.py:178]: is no one else calling this method? method behavior changing...

**Review by myprasanna — APPROVED** [review]

File: mongoengine/dereference.py [hunk 3]
```diff
@@ -36,6 +38,35 @@ def __call__(self, items, max_depth=1, instance=None, name=None):
...
+            for key in d:
+                if isinstance(d[key], dict) and d[key]:
+                    set_leaves(d[key], new_val)
+                else:
+                    d[key] = new_val
+
+        self.field_paths = {}
+        if field_paths:
+            self.max_depth = 1000
+            max_len = 0
```
Comment (Chipe1) [inline mongoengine/dereference.py:60]: is `max_len` used somewhere?

File: mongoengine/dereference.py (hunk 2 above)
Comment (squarethecircle) [reply mongoengine/dereference.py:178]: nope, internal class private method only called from this file.

File: mongoengine/dereference.py (hunk 1 above)
Comment (squarethecircle) [reply mongoengine/dereference.py:135]: need this to terminate reference fetching at the end of a path

File: mongoengine/dereference.py (hunk 3 above)
Comment (squarethecircle) [reply mongoengine/dereference.py]: whoops, no -- removed.

**Pull request merged**